rundeckapp/grails-spa/packages/ui-trellis/src/app/components/readme-motd/EditProjectFile.spec.ts

```diff
@@ -0,0 +1,153 @@
+      () => Promise.reject(new Error("Failed to fetch file")),
+    );
+    wrapper.vm.notifyError = jest.fn();
+    await wrapper.vm.getFileText();
+    expect(wrapper.vm.notifyError).toHaveBeenCalledWith("Failed to fetch file");
+  });
+
+  it("handles failure when user edits the file and fails to save it", async () => {
+    (editProjectFileService.saveProjectFile as jest.Mock).mockRejectedValue(
+      new Error("Failed to save file"),
+    );
+    wrapper.vm.notifyError = jest.fn();
+    wrapper.vm.fileText = "new content";
+    await wrapper.find('[data-test-id="save"]').trigger("click");
+    expect(wrapper.vm.notifyError).toHaveBeenCalledWith("Failed to save file");
+  });
+
+  it("handles success when user edits the file and saves it", async () => {
+    wrapper.vm.fileText = "new content";
+    await wrapper.find('[data-test-id="save"]').trigger("click");
+    expect(
+      editProjectFileService.saveProjectFile as jest.Mock,
+    ).toHaveBeenCalledWith("default", "readme.md", "new content");
+  });
+
+  it("displays admin specific message and configuration link when user is an admin", async () => {
+    const footerText = wrapper.find(".card-footer").text();
+    expect(footerText).toContain("file.warning.not.displayed.admin.message");
+    expect(wrapper.find(".card-footer a").text()).toBe(
+      "project.configuration.label",
+    );
+  });
+
+  it("displays non-admin specific message when user is not an admin", async () => {
+    await mountEditProjectFile({
+      authAdmin: false,
+    });
+    expect(wrapper.text()).toContain(
+      "file.warning.not.displayed.nonadmin.message",
+    );
+  });
+  it("does not allow non-admin user to save the file", async () => {
+    await mountEditProjectFile({ authAdmin: false });
+    wrapper.vm.save = jest.fn();
+    wrapper.vm.getFileText = jest.fn();
+
+    wrapper.find('[data-test-id="save"]').trigger("click");
+    await wrapper.vm.$nextTick();
+    expect(wrapper.vm.save).not.toHaveBeenCalled();
+  });
+  it("does not allow non-admin user to edit the file", async () => {
+    await mountEditProjectFile({ authAdmin: false });
+    expect(wrapper.find('[data-test-id="my-ace-editor"]').exists()).toBe(false);
+  });
+  it("renders the correct strings when user doesn't have permissions", async () => {
+    await mountEditProjectFile({ authAdmin: false });
+    expect(wrapper.text()).toContain(
+      "file.warning.not.displayed.nonadmin.message",
+    );
+  });
+  it("renders the correct strings when user has permissions", async () => {
+    expect(wrapper.text()).toContain(
+      "file.warning.not.displayed.admin.message",
+    );
+  });
+  it("reverts edits and triggers the correct action when the cancel button is clicked", async () => {
+    wrapper.vm.originalFileText = "sample file content";
+    wrapper.vm.fileText = "edited content";
+    wrapper.vm.createProjectHomeLink = jest.fn(() => {
+      wrapper.vm.fileText = wrapper.vm.originalFileText;
+    });
```

Comment:
When I wrote the ticket description, I was aware that if the user pressed cancel, the text wouldn't be saved and was under the impression that the component would revert back to the the original text, but turns out that upon clicking cancel, the component is triggering a navigation. 

So best thing to do is remove lines 146-148, and you trigger click on the cancel button, await the next tick and assert the document location. If the document won't be easy to test, you can replace this line in the component:

`document.location = url("project/" + this.project + "/home").href;`

to:

`window.location = url("project/" + this.project + "/home").href;`

document.location and window.location will have the same outcome, but it's easier to test the window.

And then you won't need to assert the content of fileText (line 151)

Reply:
Yes I was getting error with the [object object] and when I searched they recommended to use window.location. As I didn't want to change the component. Thanks I will do it